src/magic_folder/test/test_magic_folder_service.py

```diff
@@ -117,6 +148,182 @@ def test_create_local_snapshot(self, relative_target_path, content):
+        binary(),
+    )
+    def test_uploader_service(
+            self,
+            name,
+            file_path,
+            relative_magic_path,
+            relative_state_path,
+            author,
+            collective_dircap,
+            upload_dircap,
+            poll_interval,
+            content,
+    ):
+        """
+        ``MagicFolder.from_config`` creates an ``UploaderService`` which will
+        upload snapshots using the given Tahoe client object.
+        """
+        reactor = task.Clock()
+
+        root = create_fake_tahoe_root()
+        http_client = create_tahoe_treq_client(root)
+        tahoe_client = TahoeClient(
+            DecodedURL.from_text(U"http://example.invalid./"),
+            http_client,
+        )
+
+        root._uri.data[upload_dircap] = json.dumps([
+            u"dirnode",
+            {u"children": {}},
+        ])
+
+        root._uri.data[collective_dircap] = json.dumps([
+            u"dirnode",
+            {u"children": {}},
+        ])
+
+        basedir = FilePath(self.mktemp())
+        global_config = create_global_configuration(
+            basedir,
+            u"tcp:-1",
+            FilePath(u"/non-tahoe-directory"),
+            u"tcp:-1",
+        )
+
+        magic_path = basedir.preauthChild(relative_magic_path)
+        magic_path.asBytesMode("utf-8").makedirs()
+
+        statedir = basedir.child(u"state")
+        state_path = statedir.asTextMode("utf-8").preauthChild(relative_state_path)
+
+        target_path = magic_path.asTextMode("utf-8").preauthChild(file_path)
+        target_path.asBytesMode("utf-8").parent().makedirs(ignoreExistingDirectory=True)
+        target_path.asBytesMode("utf-8").setContent(content)
+
+        global_config.create_magic_folder(
+            name,
+            magic_path,
+            state_path,
+            author,
+            collective_dircap,
+            upload_dircap,
+            poll_interval,
+        )
+
+        magic_folder = MagicFolder.from_config(
+            reactor,
+            tahoe_client,
+            name,
+            global_config,
+        )
+
+        magic_folder.startService()
+        self.addCleanup(magic_folder.stopService)
+
+        self.assertThat(
+            magic_folder.uploader_service.running,
+            Equals(True),
+        )
+
+        self.assertThat(
+            magic_folder.uploader_service._remote_snapshot_creator._local_author,
+            Equals(author),
+        )
+
+        self.assertThat(
+            magic_folder.folder_name,
+            Equals(name),
+        )
+
+        # add a file. This won't actually add a file until we advance
+        # the clock.
+        d = magic_folder.local_snapshot_service.add_file(
+            target_path,
+        )
+
+        self.assertThat(
+            d,
+            succeeded(Always()),
+        )
+
+        def children():
+            return json.loads(root._uri.data[upload_dircap])[1][u"children"]
+
```

Comment:
I guess *ideally* this would be confirmed via the API instead of internals .. that is, by calling something on `tahoe_client` like `TahoeClient.list_directory(upload_dircap`).

So I think that could be like `children_d = TahoeClient.list_directory(upload_dircap)` and then `self.assertThat(children_d, succeeded(AfterPreprocessing(result=Equals({})))` or similar?

Reply:
Maybe ... on the other hand, this is a test for code that does uploads.  If there is a coherent model underlying the HTTP interface then I don't think there's a compelling reason to force the code that checks to see if the upload has happened properly to also go via HTTP.  It is possible to argue that `json.loads(root._uri.data[upload_dircap])[1][u"children"]` is not the *most* coherent underlying model though ...